Integrate torchgen exception boundary with ExecuTorch

swolchok opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #7042
* #7041
* __->__ #7546
* #7040

As of #7746, we build with exceptions by default, so we just need to use them.

Differential Revision: [D67904052](https://our.internmc.facebook.com/intern/diff/D67904052/)

swolchok committed:
Build optimized operators lib with -fexceptions

It should be permissible for this library to use exceptions, since it is not required to supported embedded systems.

TODO: presumably we need to manage rollout of the torchgen patch?

Differential Revision: [D67904052](https://our.internmc.facebook.com/intern/diff/D67904052/)

[ghstack-poisoned]

swolchok committed:
Update on "Build optimized operators lib with -fexceptions"


It should be permissible for this library to use exceptions, since it is not required to supported embedded systems.

TODO: presumably we need to manage rollout of the torchgen patch?

Differential Revision: [D67904052](https://our.internmc.facebook.com/intern/diff/D67904052/)

[ghstack-poisoned]

swolchok commented:
converting to draft to prevent accidental merge until the PR is updated to include a proper PyTorch pin bump; right now it points directly at https://github.com/pytorch/pytorch/pull/144341 . It will still be reviewable whenever I get CI fixed

huydhn commented on build/Codegen.cmake:
```diff
@@ -11,13 +11,14 @@
   message(STATUS "  OPS_SCHEMA_YAML: ${GEN_OPS_SCHEMA_YAML}")
   message(STATUS "  ROOT_OPS: ${GEN_ROOT_OPS}")
   message(STATUS "  INCLUDE_ALL_OPS: ${GEN_INCLUDE_ALL_OPS}")
+  message(STATUS "  ADD_EXCEPTION_BOUNDARY: ${GEN_ADD_EXCEPTION_BOUNDARTY}")
```
Small typo

```suggestion
  message(STATUS "  ADD_EXCEPTION_BOUNDARY: ${ADD_EXCEPTION_BOUNDARY}")
```

huydhn commented on build/Codegen.cmake:
```diff
@@ -93,7 +96,10 @@ function(generate_bindings_for_kernels)
-  )
+    )
+  if(GEN_ADD_EXCEPTION_BOUNDARY)
+    set(_gen_command "${_gen_command} --add-exception-boundary")
```
I think the set command use semi-colon separator instead of space, i.e. `/opt/conda/envs/py_3.10/bin/python3;-m;torchgen.gen_executorch;--source-path=/pytorch/executorch/configurations/../codegen;--install-dir=/pytorch/executorch/pip-out/temp.linux-x86_64-cpython-310/cmake-out/configurations/optimized_native_cpu_ops_lib;--tags-path=/opt/conda/envs/py_3.10/lib/python3.10/site-packages/torchgen/packaged/ATen/native/tags.yaml;--aten-yaml-path=/opt/conda/envs/py_3.10/lib/python3.10/site-packages/torchgen/packaged/ATen/native/native_functions.yaml;--op-selection-yaml-path=/pytorch/executorch/pip-out/temp.linux-x86_64-cpython-310/cmake-out/configurations/optimized_native_cpu_ops_lib/selected_operators.yaml;--add-exception-boundary`

```suggestion
    set(_gen_command "${_gen_command}" --add-exception-boundary)
```

swolchok replied:
strange, I thought I built this locally. Giving it a shot now.

swolchok replied:
this does seem to have improved the situation (thanks!) but the torchgen version still seems to be wrong: https://github.com/pytorch/executorch/actions/runs/12680876952/job/35343491373?pr=7546#step:9:2186

swolchok replied:
@huydhn still seeing at least some wrong-torchgen-version failures:
- [test-arm-backend-delegation](https://github.com/pytorch/executorch/actions/runs/12699647767/job/35400710061?pr=7546)
- [test-arm-reference-delegation](https://github.com/pytorch/executorch/actions/runs/12699647767/job/35400706527?pr=7546#step:14:1397)
- [test-llava-runner-linux](https://github.com/pytorch/executorch/actions/runs/12699647761/job/35400719004?pr=7546#step:14:8748)
- [test-phi3-mini-runner-linux](https://github.com/pytorch/executorch/actions/runs/12699647761/job/35400719596?pr=7546#step:14:8614)
- [test-eval_llama-wikitext-linux](https://github.com/pytorch/executorch/actions/runs/12699647761/job/35400718418?pr=7546#step:14:8658)
- [test-eval_llama-mmlu-linux](https://github.com/pytorch/executorch/actions/runs/12699647761/job/35400717012?pr=7546#step:14:8644)
- [test-llama_runner_eager-linux](https://github.com/pytorch/executorch/actions/runs/12699647761/job/35400717909?pr=7546#step:14:8625)

however, we do have mac successes. looking into the common thread for these failures now.

swolchok replied:
the ARM one is easy: the workflow directly invokes `install_executorch` from .ci/scripts/utils.sh without "use-pt-pinned-commit". Is there a reason "use-pt-pinned-commit" isn't the default or only option?

test-llava-runner-linux reruns install_requirements.sh and doesn't pass --use-pt-pinned-commit, same with test-phi-3-mini-runner, test-eval_llama-wikitext-linux, test-eval_llama-mmlu-linux, and test-llama_runner_eager-linux. I will send a PR to fix those.

swolchok committed:
Update on "Build optimized operators lib with -fexceptions"


It should be permissible for this library to use exceptions, since it is not required to supported embedded systems.

TODO: presumably we need to manage rollout of the torchgen patch?

Differential Revision: [D67904052](https://our.internmc.facebook.com/intern/diff/D67904052/)

[ghstack-poisoned]

swolchok committed:
Update on "Build optimized operators lib with -fexceptions"


It should be permissible for this library to use exceptions, since it is not required to supported embedded systems.

TODO: presumably we need to manage rollout of the torchgen patch?

Differential Revision: [D67904052](https://our.internmc.facebook.com/intern/diff/D67904052/)

[ghstack-poisoned]

swolchok committed:
Update on "Build optimized operators lib with -fexceptions"


It should be permissible for this library to use exceptions, since it is not required to supported embedded systems.

TODO: presumably we need to manage rollout of the torchgen patch?

Differential Revision: [D67904052](https://our.internmc.facebook.com/intern/diff/D67904052/)

[ghstack-poisoned]

swolchok committed:
Update on "Build optimized operators lib with -fexceptions"


It should be permissible for this library to use exceptions, since it is not required to supported embedded systems.

TODO: presumably we need to manage rollout of the torchgen patch?

Differential Revision: [D67904052](https://our.internmc.facebook.com/intern/diff/D67904052/)

[ghstack-poisoned]

Commit:
Update on "Build optimized operators lib with -fexceptions"


It should be permissible for this library to use exceptions, since it is not required to supported embedded systems.

TODO: presumably we need to manage rollout of the torchgen patch?

Differential Revision: [D67904052](https://our.internmc.facebook.com/intern/diff/D67904052/)

[ghstack-poisoned]

Commit:
Update on "Build optimized operators lib with -fexceptions"


It should be permissible for this library to use exceptions, since it is not required to supported embedded systems.

TODO: presumably we need to manage rollout of the torchgen patch?

Differential Revision: [D67904052](https://our.internmc.facebook.com/intern/diff/D67904052/)

[ghstack-poisoned]

Commit:
Update on "Build optimized operators lib with -fexceptions"


It should be permissible for this library to use exceptions, since it is not required to supported embedded systems.

TODO: presumably we need to manage rollout of the torchgen patch?

Differential Revision: [D67904052](https://our.internmc.facebook.com/intern/diff/D67904052/)

[ghstack-poisoned]

Commit:
Update on "Build optimized operators lib with -fexceptions"


It should be permissible for this library to use exceptions, since it is not required to supported embedded systems.

TODO: presumably we need to manage rollout of the torchgen patch?

Differential Revision: [D67904052](https://our.internmc.facebook.com/intern/diff/D67904052/)

[ghstack-poisoned]

Commit:
Update on "Build optimized operators lib with -fexceptions"


It should be permissible for this library to use exceptions, since it is not required to supported embedded systems.

TODO: presumably we need to manage rollout of the torchgen patch?

Differential Revision: [D67904052](https://our.internmc.facebook.com/intern/diff/D67904052/)

[ghstack-poisoned]

Commit:
Update on "Build optimized operators lib with -fexceptions"


It should be permissible for this library to use exceptions, since it is not required to supported embedded systems.

TODO: presumably we need to manage rollout of the torchgen patch?

Differential Revision: [D67904052](https://our.internmc.facebook.com/intern/diff/D67904052/)

[ghstack-poisoned]

swolchok committed:
Update on "Build optimized operators lib with -fexceptions"


It should be permissible for this library to use exceptions, since it is not required to supported embedded systems.

TODO: presumably we need to manage rollout of the torchgen patch?

Differential Revision: [D67904052](https://our.internmc.facebook.com/intern/diff/D67904052/)

[ghstack-poisoned]

swolchok committed:
Update on "Build optimized operators lib with -fexceptions"


It should be permissible for this library to use exceptions, since it is not required to supported embedded systems.

TODO: presumably we need to manage rollout of the torchgen patch?

Differential Revision: [D67904052](https://our.internmc.facebook.com/intern/diff/D67904052/)

[ghstack-poisoned]

swolchok commented:
This isn't quite right. -fno-exceptions (and, to a lesser extent, -fno-rtti) should be all-or-none within a single static library or binary, otherwise one runs the risk of surprising behavior similar to that exhibited by One Definition Rule (ODR) violations. Instead, we need a mechanism that makes sure we build everything with -fexceptions if we are including optimized kernels, extensions, or any other component that could share code with PyTorch; we can only use `-fno-exceptions` if we're excluding all of that.

swolchok commented:
https://github.com/pytorch/executorch/discussions/7736 should determine how to replace this

swolchok committed:
Update on "Build optimized operators lib with -fexceptions"


It should be permissible for this library to use exceptions, since it is not required to support embedded systems.

TODO: presumably we need to manage rollout of the torchgen patch?

Differential Revision: [D67904052](https://our.internmc.facebook.com/intern/diff/D67904052/)

[ghstack-poisoned]

swolchok committed:
Update on "Build optimized operators lib with -fexceptions"


It should be permissible for this library to use exceptions, since it is not required to support embedded systems.

TODO: presumably we need to manage rollout of the torchgen patch?

Differential Revision: [D67904052](https://our.internmc.facebook.com/intern/diff/D67904052/)

[ghstack-poisoned]

swolchok committed:
Update on "Build optimized operators lib with -fexceptions"


It should be permissible for this library to use exceptions, since it is not required to support embedded systems.

TODO: presumably we need to manage rollout of the torchgen patch?

Differential Revision: [D67904052](https://our.internmc.facebook.com/intern/diff/D67904052/)

[ghstack-poisoned]

swolchok committed:
Update on "Integrate torchgen exception boundary with ExecuTorch"


As of #7746, we build with exceptions by default, so we just need to use them.

TODO: presumably we need to manage rollout of the torchgen patch?

Differential Revision: [D67904052](https://our.internmc.facebook.com/intern/diff/D67904052/)

[ghstack-poisoned]

swolchok committed:
Update on "Integrate torchgen exception boundary with ExecuTorch"


As of #7746, we build with exceptions by default, so we just need to use them.

TODO: presumably we need to manage rollout of the torchgen patch?

Differential Revision: [D67904052](https://our.internmc.facebook.com/intern/diff/D67904052/)

[ghstack-poisoned]

swolchok committed:
Update on "Integrate torchgen exception boundary with ExecuTorch"


As of #7746, we build with exceptions by default, so we just need to use them.

TODO: presumably we need to manage rollout of the torchgen patch?

Differential Revision: [D67904052](https://our.internmc.facebook.com/intern/diff/D67904052/)

[ghstack-poisoned]

swolchok commented:
looks like tests are good; the github export mechanism just doesn't seem to be basing the PRs off main

mergennachin commented:
@swolchok - FYI, there's a PR here to improve torchgen package import (https://github.com/pytorch/executorch/pull/6730)

Somewhat complementary but related

Commit:
Update on "Integrate torchgen exception boundary with ExecuTorch"


As of #7746, we build with exceptions by default, so we just need to use them.

Differential Revision: [D67904052](https://our.internmc.facebook.com/intern/diff/D67904052/)

[ghstack-poisoned]

Commit:
Update on "Integrate torchgen exception boundary with ExecuTorch"


As of #7746, we build with exceptions by default, so we just need to use them.

Differential Revision: [D67904052](https://our.internmc.facebook.com/intern/diff/D67904052/)

[ghstack-poisoned]

Commit:
Update on "Integrate torchgen exception boundary with ExecuTorch"


As of #7746, we build with exceptions by default, so we just need to use them.

Differential Revision: [D67904052](https://our.internmc.facebook.com/intern/diff/D67904052/)

[ghstack-poisoned]

Commit:
Update on "Integrate torchgen exception boundary with ExecuTorch"


As of #7746, we build with exceptions by default, so we just need to use them.

Differential Revision: [D67904052](https://our.internmc.facebook.com/intern/diff/D67904052/)

[ghstack-poisoned]

Commit:
Update on "Integrate torchgen exception boundary with ExecuTorch"


As of #7746, we build with exceptions by default, so we just need to use them.

Differential Revision: [D67904052](https://our.internmc.facebook.com/intern/diff/D67904052/)

[ghstack-poisoned]
